state/src/account.rs

```diff
@@ -0,0 +1,401 @@
+        })
+    }
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum CodeState {
+    Clean,
+    Dirty,
+}
+
+#[derive(Debug)]
+pub struct StateObject {
+    balance: U256,
+    nonce: U256,
+    storage_root: H256,
+    code_hash: H256,
+    code: Vec<u8>,
+    code_size: usize,
+    code_state: CodeState,
+    storage_changes: HashMap<H256, H256>,
+}
+
+impl From<Account> for StateObject {
+    fn from(account: Account) -> Self {
+        StateObject {
+            balance: account.balance,
+            nonce: account.nonce,
+            storage_root: account.storage_root,
+            code_hash: account.code_hash,
+            code: vec![],
+            code_size: 0,
+            code_state: CodeState::Clean,
+            storage_changes: HashMap::new(),
+        }
+    }
+}
+
+impl StateObject {
+    /// Create a new account.
+    /// NOTE: If contract account generated, make sure you use `init_code` on
+    /// this before `commit`ing.
+    pub fn new(balance: U256, nonce: U256) -> StateObject {
+        StateObject {
+            balance,
+            nonce,
+            storage_root: SHA3_NULL_RLP,
+            code_hash: SHA3_EMPTY,
+            code: vec![],
+            code_size: 0,
+            code_state: CodeState::Clean,
+            storage_changes: HashMap::new(),
+        }
+    }
+
+    /// Create a new account from rlp bytes.
+    /// Note: make sure you use `read_code` after this.
+    pub fn from_rlp(data: &[u8]) -> Result<StateObject, Error> {
+        let account: Account = rlp::decode(data)?;
+        Ok(account.into())
+    }
+
+    pub fn rlp(&self) -> Vec<u8> {
+        rlp::encode(&self.account())
+    }
+
+    pub fn account(&self) -> Account {
+        Account {
+            balance: self.balance,
+            nonce: self.nonce,
+            storage_root: self.storage_root,
+            code_hash: self.code_hash,
+        }
+    }
+
+    pub fn init_code(&mut self, code: Vec<u8>) {
+        self.code = code.clone();
+        self.code_hash = From::from(&Sha3_256::digest(&code)[..]);
+        self.code_size = code.len();
+        self.code_state = CodeState::Dirty;
+    }
+
+    pub fn read_code<B: DB>(&mut self, db: &mut B) -> Result<Vec<u8>, Error> {
+        if self.code_hash == SHA3_EMPTY {
+            return Ok(vec![]);
+        }
+        if !self.code.is_empty() {
+            return Ok(self.code.clone());
+        }
+        let c = db
+            .get(&self.code_hash)
+            .or_else(|e| Err(Error::DB(format!("{}", e))))?
+            .unwrap_or_else(|| vec![]);
+        self.code = c.clone();
+        self.code_size = c.len();
+        self.code_state = CodeState::Clean;
+        Ok(c)
+    }
+
+    pub fn balance(&self) -> U256 {
+        self.balance
+    }
+
+    pub fn nonce(&self) -> U256 {
+        self.nonce
+    }
+
+    pub fn code(&self) -> Option<Vec<u8>> {
+        if self.code.is_empty() {
+            return None;
+        }
+        Some(self.code.clone())
```

Comment:
What about just return self.code? The empty code and None are indistinguishable.

Reply:
Good idea. Seems I counld use Vec.is_empty instead of match Option.